File: frontend/src/types/interview.ts
// Interview module TypeScript types

export interface InterviewQuestion {
    id: string;
    topic: string;
    question: string;
    difficulty: 'easy' | 'medium' | 'hard';
    source?: string;
    user_answer?: string;
    ai_score?: string;
    ai_verdict?: string;
    ai_feedback?: string;
    model_answer?: string;
    created_at?: string;
}

export interface InterviewSession {
    id: string;
    branch: string;
    topic?: string;
    status: 'active' | 'completed' | 'abandoned';
    created_at?: string;
    questions: InterviewQuestion[];
}

export interface InterviewSessionSummary {
    id: string;
    branch: string;
    topic?: string;
    status: 'active' | 'completed' | 'abandoned';
    created_at?: string;
    question_count: number;
}

export interface GeneratedQuestionsResponse {
    student_id: string;
    branch: string;
    semester: number;
    overall_gpa: string;
    weak_subjects: string[];
    questions: InterviewQuestion[];
    /** 'built-in' | 'groq_jd' | 'groq_resume' | 'groq_jd_resume' | *_gemini variants */
    source: string;
}

export interface AnswerSubmitResponse {
    id: string;
    question: string;
    user_answer: string;
    session_completed: boolean;
}

export interface QuestionEvaluation {
    question_id: string;
    question: string;
    topic: string;
    difficulty: string;
    user_answer: string;
    ai_score: number;
    ai_verdict: 'Strong' | 'Adequate' | 'Weak';
    ai_feedback: string;
    model_answer: string;
}

export interface EvaluationResult {
    session_id: string;
    total_questions: number;
    avg_score: number;
    strong_count: number;
    adequate_count: number;
    weak_count: number;
    overall_verdict: string;
    questions: QuestionEvaluation[];
}

File: backend/app/schemas/interview.py
"""
Interview Pydantic Schemas (Phase 2)
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


# ---------------------------------------------------------------------------
# Question schemas
# ---------------------------------------------------------------------------

class InterviewQuestionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    topic: str
    question: str
    difficulty: str
    source: Optional[str] = None
    follow_up: Optional[str] = None
    user_answer: Optional[str] = None
    ai_score: Optional[int] = None
    ai_verdict: Optional[str] = None
    ai_feedback: Optional[str] = None
    model_answer: Optional[str] = None
    mistakes: Optional[list] = None
    improvement: Optional[str] = None
    created_at: Optional[datetime] = None


class AnswerSubmitRequest(BaseModel):
    question_id: uuid.UUID
    answer: str


class AnswerSubmitResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    question: str
    user_answer: str
    session_completed: bool = False  # True when all questions in the session answered


# ---------------------------------------------------------------------------
# Session schemas
# ---------------------------------------------------------------------------

class InterviewSessionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    branch: str
    topic: Optional[str] = None
    status: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None  # advances when status changes (e.g. → completed)
    questions: list[InterviewQuestionOut] = []


# ---------------------------------------------------------------------------
# Question generation response
# ---------------------------------------------------------------------------

class GeneratedQuestionsResponse(BaseModel):
    student_id: str
    branch: str
    semester: int
    overall_gpa: str
    weak_subjects: list[str]
    questions: list[dict]
    source: str = "built-in"   # "built-in" | "groq_jd" | "groq_resume" | "groq_jd_resume" | *_gemini variants


# ---------------------------------------------------------------------------
# Session create request (used by POST /sessions)
# ---------------------------------------------------------------------------

class SessionCreateRequest(BaseModel):
    jd_text: str = ""           # Job description text (optional when resume is provided)
    resume_context: Optional[str] = None
    limit: int = 10
